### PR TITLE
Symmetry post-selection

### PR DESCRIPTION
**Pull request by KrzysztofB-1qbit**
Adds the symmetry post-selection functionality
- Adds ancilla_symmetry_circuit function to append the symmetry verification circuit to a given circuit
- Adds post_select function to pst select on a dictionary of ancilla qubits
- Adds a convenience strip_post_selection function to remove the ancilla and aggregate the data as if no post-selection was used
- Adds tests for the functions

### CODE REVIEW

File: tangelo/toolboxes/post_processing/post_selection.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+            and entanglement with an ancilla qubit"""
+    n_qubits = circuit.width
+    basis_gates = measurement_basis_gates(pauli_string_to_of(sym_op))
+    basis_circ = Circuit(basis_gates)
+    parity_gates = [Gate("CNOT", n_qubits, i) for i in range(n_qubits)]
+    circuit_new = circuit + basis_circ + Circuit(parity_gates) + basis_circ.inverse()
+    return circuit_new
+
+
+def post_select(hist, expected_outcomes):
```
Comment (AlexandreF-1qbit) [inline tangelo/toolboxes/post_processing/post_selection.py:39]: Is there a reason we want a post_select function taking frequencies and ignoring the number of shots? If not, we could only reuse the `Histogram.post_select` function and delete this function.

**Review by AlexandreF-1qbit — CHANGES_REQUESTED** [review]
I feel some functions can be rewritten with the Histogram class and an arbitrary number of shots (I think specifying `n_shots=1` would work).

File: tangelo/toolboxes/post_processing/post_selection.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+from tangelo.linq.helpers import measurement_basis_gates, pauli_string_to_of
+
+
+def ancilla_symmetry_circuit(circuit, sym_op):
+    """Appends to a circuit a symmetry operator and ancilla for symmetry verification.
+    For more details see arXiv:1807.10050.
+
+    Args:
+        circuit (Circuit): A quantum circuit to be equipped with symmetry verification
+        sym_op (string): The symmetry operator to be verified, in the form of a Pauli string
```
[the remark this answers is not in the document]
Comment (ValentinS4t1qbit) [reply tangelo/toolboxes/post_processing/post_selection.py:26]: Not fond of the string being what we expect from the user. As Alexandre said, we tend to manipulate `QubitOperator` objects, or openfermion-style (tuple) representations (which causes you to use `pauli_string_to_of`).

I'm not sure why we'd support the string case, I'd need to be convinced. I think for consistency with the rest of the package, we should not.

File: tangelo/toolboxes/post_processing/post_selection.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""This module provides functions to create symmetry verification and postselection circuits."""
+from tangelo.linq import Circuit, Gate
+from tangelo.linq.helpers import measurement_basis_gates, pauli_string_to_of
+
+
+def ancilla_symmetry_circuit(circuit, sym_op):
+    """Appends to a circuit a symmetry operator and ancilla for symmetry verification.
```
[the remark this answers is not in the document]
Comment (ValentinS4t1qbit) [reply tangelo/toolboxes/post_processing/post_selection.py:21]: How about we say

`"""Append a symmetry operator circuit using an extra ancilla qubit, for the purpose of symmetry verification, to the input circuit."""`

File: tangelo/toolboxes/post_processing/post_selection.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    n_qubits = circuit.width
+    basis_gates = measurement_basis_gates(pauli_string_to_of(sym_op))
+    basis_circ = Circuit(basis_gates)
+    parity_gates = [Gate("CNOT", n_qubits, i) for i in range(n_qubits)]
+    circuit_new = circuit + basis_circ + Circuit(parity_gates) + basis_circ.inverse()
+    return circuit_new
+
+
+def post_select(hist, expected_outcomes):
+    """Applies post selection to frequency data based on
```
Comment (ValentinS4t1qbit) [inline tangelo/toolboxes/post_processing/post_selection.py:40]: The custom is not to conjugate the verbs in docstrings and comments, when you start with a verb like that.
`Apply post-selection to....`

File: tangelo/toolboxes/post_processing/post_selection.py (hunk 1 above)
Comment (ValentinS4t1qbit) [reply tangelo/toolboxes/post_processing/post_selection.py:39]: Whatever we do, let's avoid defining redundant code in several places and maybe try to write one functionality using the other, or as a special case or the other.

But first and foremost: let's clarify what this new functionality brings, that we did not have before.

**Review by ValentinS4t1qbit — CHANGES_REQUESTED** [review]
We need to figure out if some of these new functions implement a useful feature. If they do, how can be reuse similar existing code to avoid redundancy (e.g can we implement a functionality by reusing code from the other, or expressing it as a special case of the other).

Note: there are no tests for `ancilla_symmetry_circuit`. Are there any corner cases or tricky situations we can think of, that would warrant writing a test ?

If not, at the very least you should have tried an input circuit, shoved it into that function and drawn the output to see that it was what you expected.

File: tangelo/toolboxes/post_processing/post_selection.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (KrzysztofB-1qbit) [reply tangelo/toolboxes/post_processing/post_selection.py:26]: I was thinking it might be more intuitive for the user to use Pauli strings. This is more like a human readable flag which would normally be used "by hand", where the user takes a circuit and decides which symmetry to input. It's much easier for a human to say "XX" than [(0, 'X'), (1, 'X')] or construct a whole QubitOperator. I can add functionality to work with other types just in case someone needs to do that though. 

Also, could add some consistency checks to make sure the number of qubits matches the input circuit.

File: tangelo/toolboxes/post_processing/post_selection.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+    # Check if the operator size matches the circuit width
+    if n_qubits < op_len:
+        raise RuntimeError("The size of the symmetry operator is bigger than the circuit width.")
+    elif n_qubits > op_len:
+        warnings.warn("The size of the symmetry operator is smaller than the circuit width. Remaining qubits will be measured in the Z-basis.")
+
+    if isinstance(sym_op, str):
+        basis_gates = measurement_basis_gates(pauli_string_to_of(sym_op))
+    elif isinstance(sym_op, list or tuple):
```
Comment (ValentinS4t1qbit) [inline tangelo/toolboxes/post_processing/post_selection.py:53]: I've never seen this for testing several types, and I don't know what it actually does.
I recommend `(list, tuple)` as your second argument, as it works for sure and is commonly recommended: https://stackoverflow.com/questions/33311258/python-check-if-variable-isinstance-of-any-type-in-list

File: tangelo/toolboxes/post_processing/post_selection.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+            and entanglement with an ancilla qubit, which is added as the last qubit.
+            Increases the input circuit width by 1."""
+    if isinstance(sym_op, QubitOperator):
+        op_len = count_qubits(sym_op)
+    else:
+        op_len = len(sym_op)
+    n_qubits = circuit.width
+
+    # Check if the operator size matches the circuit width
+    if n_qubits < op_len:
```
Comment (ValentinS4t1qbit) [inline tangelo/toolboxes/post_processing/post_selection.py:46]: In both your error or warning messages, I think we can talk about number of qubits instead of width.

File: tangelo/toolboxes/post_processing/post_selection.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        raise RuntimeError("The size of the symmetry operator is bigger than the circuit width.")
+    elif n_qubits > op_len:
+        warnings.warn("The size of the symmetry operator is smaller than the circuit width. Remaining qubits will be measured in the Z-basis.")
+
+    if isinstance(sym_op, str):
```
Comment (ValentinS4t1qbit) [inline tangelo/toolboxes/post_processing/post_selection.py:51]: You have no `else` here. I don't think it's something you've done in this function: if the `symptoms_op` is none of these 3 things then we should return an error. That should cover our backs if the user pushes in something of the wrong type.

**Review by ValentinS4t1qbit — CHANGES_REQUESTED** [review]
Some very minor things that should be straightforward. Almost there !

File: tangelo/toolboxes/post_processing/post_selection.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+    Args:
+        circuit (Circuit): A quantum circuit to be equipped with symmetry verification
+        sym_op (string): The symmetry operator to be verified.
+            Can be a Pauli string, OpenFermion style list or a QubitOperator
+
+    Returns:
+        Circuit: The input circuit appended with the proper basis rotation
+            and entanglement with an ancilla qubit, which is added as the last qubit.
+            Increases the input circuit width by 1."""
```
[the remark this answers is not in the document]
Comment (KrzysztofB-1qbit) [reply tangelo/toolboxes/post_processing/post_selection.py]: ok

File: tangelo/toolboxes/post_processing/post_selection.py (hunk 5 above)
Comment (KrzysztofB-1qbit) [reply tangelo/toolboxes/post_processing/post_selection.py]: ok

File: tangelo/toolboxes/post_processing/post_selection.py (hunk 6 above)
Comment (KrzysztofB-1qbit) [reply tangelo/toolboxes/post_processing/post_selection.py:47]: changed

File: tangelo/toolboxes/post_processing/post_selection.py (hunk 7 above)
Comment (KrzysztofB-1qbit) [reply tangelo/toolboxes/post_processing/post_selection.py:52]: Added error message

**Review by ValentinS4t1qbit — APPROVED** [review]
Thank you Krzys :)

**Review by AlexandreF-1qbit — APPROVED** [review]
